src/metadata/metadata_passive_io.c

```diff
@@ -0,0 +1,377 @@
+		uint64_t raw_end_page = raw_start_page + raw->ssd_pages;
+		uint64_t overlap_start = OCF_MAX(io_start_page, raw_start_page);
+		uint64_t overlap_end = OCF_MIN(io_end_page, raw_end_page);
+		uint64_t overlap_start_data = overlap_start - io_start_page;
+		uint64_t overlap_page;
+		uint64_t overlap_count;
+
+		if (overlap_start >= overlap_end)
+			continue;
+
+		overlap_page = overlap_start - raw_start_page;
+		overlap_count = overlap_end - overlap_start;
+
+		if (seg == metadata_segment_collision) {
+			/* The collision is not updated yet but the range of affected
+			   cachelines is already known. Remove the old mapping related info
+			   from the metadata */
+			/* The range of cachelines with updated collision section */
+			cache_line_range_start = overlap_page * raw->entries_in_page;
+			cache_line_range_end = cache_line_range_start +
+				raw->entries_in_page * overlap_count;
+
+			cleanup_old_mapping(cache, cache_line_range_start,
+					cache_line_range_end);
+		}
+
+		ctx_data_seek(cache->owner, data, ctx_data_seek_begin,
+				PAGES_TO_BYTES(overlap_start_data));
+		ocf_metadata_raw_update(cache, raw, data, overlap_page, overlap_count);
+
+		if (seg != metadata_segment_collision)
+			continue;
+
+		update_list_segment(cache, cache_line_range_start,
+				cache_line_range_end);
+	}
+
+	ocf_pio_async_unlock(req->cache->standby.concurrency, req);
+	io_cmpl(io, 0);
+	env_allocator_del(cache->standby.allocator, req);
+	return 0;
+}
+
+static struct ocf_io_if passive_io_restart_if = {
+	.read = passive_io_resume,
+	.write = passive_io_resume,
+};
+
+static void passive_io_page_lock_acquired(struct ocf_request *req)
+{
+	ocf_engine_push_req_front(req, true);
+}
+
+int ocf_metadata_passive_update(ocf_cache_t cache, struct ocf_io *io,
+		ocf_end_io_t io_cmpl)
+{
+	struct ocf_metadata_ctrl *ctrl = cache->metadata.priv;
+	struct ocf_request *req;
+	uint64_t io_start_page = BYTES_TO_PAGES(io->addr);
+	uint64_t io_end_page = io_start_page + BYTES_TO_PAGES(io->bytes);
+	int lock = 0;
+
+	if (io->dir == OCF_READ) {
+		io_cmpl(io, 0);
+		return 0;
+	}
+
+	if (io_start_page >= ctrl->count_pages) {
+		io_cmpl(io, 0);
+		return 0;
+	}
+
+	if (io->addr % PAGE_SIZE || io->bytes % PAGE_SIZE) {
+		ocf_cache_log(cache, log_crit,
+				"Metadata update not aligned to page size!\n");
+		io_cmpl(io, -OCF_ERR_INVAL);
+		return -OCF_ERR_INVAL;
+	}
+
+	if (io->bytes > MAX_PASSIVE_IO_SIZE) {
+		//FIXME handle greater IOs
+		ocf_cache_log(cache, log_crit,
+				"IO size exceedes max supported size!\n");
+		io_cmpl(io, -OCF_ERR_INVAL);
+		return -OCF_ERR_INVAL;
+	}
+
+	req = (struct ocf_request*)env_allocator_new(cache->standby.allocator);
+	if (!req) {
+		io_cmpl(io, -OCF_ERR_NO_MEM);
+		return -OCF_ERR_NO_MEM;
+	}
+
+	req->io_queue = io->io_queue;;
+	req->info.internal = true;
+	req->io_if = &passive_io_restart_if;
+	req->rw = OCF_WRITE;
+	req->data = io;
+	req->master_io_req = io_cmpl;
+	req->cache = cache;
+	env_atomic_set(&req->lock_remaining, 0);
```

Comment:
extracting generic part (required for alock) from ocf_request structure and creating dedicated passive request struct would make this look way better